### ISSUE TITLE
Is try-catch scoping as hard as let and for?

### ISSUE THREAD

**Issue by tkf** [JuliaLang/julia#34668]
Given that `let x` and `for x in ...` introduces a hard-scope for `x`, I find that it's inconsistent that the scoping of `try ... catch err; ... end` is ambiguous.

But I couldn't follow all the comments in #33864 and related issues so I may be missing something.  I only found that it was explicitly decided that `let` uses a hard-scope https://github.com/JuliaLang/julia/pull/33864#issuecomment-575816436.

MWE:

```julia
julia> VERSION
v"1.5.0-DEV.230"

julia> include_string(Main, """
       x = let x
           x = 1
       end

       for x in 1:1
           x = 1
       end
       """)  # no warning

julia> include_string(Main, """
       err = try
           error("hello")
       catch err
           err
       end
       """)
┌ Warning: Assignment to `err` in soft scope is ambiguous because a global variable by the same name exists: `err` will be treated as a new local. Disambiguate by using `local err` to suppress this warning or `global err` to assign to the existing global variable.
└ @ string:1
ErrorException("hello")
```

**Comment by StefanKarpinski** [comment]
`for` does not introduce a hard scope, only `let` and functions do. The loop variable in a for loop is always local, however.

**Comment by tkf** [comment]
That's what I meant to say. Shouldn't the exception variable (`err` in `try ... catch err; ... end`) be always local?

**Comment by StefanKarpinski** [comment]
I suppose that does seem quite similar to the loop variable in a for loop.

**Comment by JeffBezanson** [comment]
Yes, I agree the catch block var should be let-bound. I was surprised to see (this really seems like a mistake of some sort) that we just lower this to a plain assignment. So for example
```
julia> function f()
         local x
         try
           error(2)
         catch x
         end
         x
       end
f (generic function with 1 method)

julia> f()
ErrorException("2")
```
That can't have been what we intended, but technically this is a breaking change.

**Comment by tkf** [comment]
Actually, I sometimes write code that relies on it. Something like

```julia
@testcase begin
    err = nothing
    @test try
        f()
        false
    catch err
        true
    end
    @test occursin("some error message", sprint(showerror, err))
end
```

I thought it was weird but I didn't think carefully back then when I started using it.  (Before this I was using `err = try ... catch err; err; end` pattern.)

I personally prefer "tightening" the scope (so breaking the above code) but I'd imagine someone else might be using it already due to Hyrum's law.  But if it is only me by some kind of miracle, please go ahead and change it :)

**Comment by StefanKarpinski** [comment]
Definitely worth a PkgEval run to see if it breaks much. Hopefully not!
